### ISSUE TITLE
resolution calculation for EnergyDispersion does not work with non-squared edisp matrices

### ISSUE THREAD

**Issue by cosimoNigro** [gammapy/gammapy#1132]
Hi all,
when you try to calculate the resolution with `get_resolution` from the `EnergyDispersion` class you will encounter this error if you have a non-sqaured matrix
```python   
    temp = (self.e_reco._interp_nodes() - mean) ** 2
ValueError: operands could not be broadcast together with shapes (72,) (108,) 
```

I think the problem is that `mean` - obtained with `_get_mean` - has the size of the `e_true` array while 
`self.e_reco._interp_nodes()` has the interpolation nodes, i.e. the dimension of `e_reco`.

Could you please, for my understanding, comment on this approach of the bias and resolution as first and second moment of the distribution?
What is done in MAGIC is to fit a gaussian in each `e_true` bin: its mean will be the bias and its sigma the dispersion. Of course this approach is valid in case of binned data, while conflicts with your approach of choosing an interpolated pdf.

Anyhow here the code to reproduce the issue. 

Thanks a lot

```python
# coding: utf-8

# issue with the energy binning
# author:
# * Cosimo Nigro (cosimonigro2@gmail.com)

import numpy as np
import astropy
import regions
import sherpa
import os

# units and coordinates from astropy
import astropy.units as u
from astropy.io import fits
from astropy.coordinates import SkyCoord, Angle
from astropy.table import Table
# regions 
from regions import CircleSkyRegion
# Datastore is the interface to the index of observations 
import gammapy.data as data # too many functions to import
# The background method
import gammapy.background as background
# Some utility function to deal with energy axis (log binning with a unit etc)
from gammapy.utils.energy import EnergyBounds
# The class performing the extraction and extraction results
from gammapy.spectrum import SpectrumExtraction, SpectrumObservation, SpectrumFit, SpectrumResult
# Spectral models
from gammapy.spectrum.models import LogParabola
# Utilities for flux points plotting
from gammapy.spectrum import FluxPoints, SpectrumEnergyGroupMaker, FluxPointEstimator
# A class to deal with exclusion regions
from gammapy.image import SkyImage
from gammapy.extern.pathlib import Path

import matplotlib.pyplot as plt


datastore = data.DataStore.from_dir('$GAMMAPY_EXTRA/datasets/hess-crab4-hd-hap-prod2/')
obs_ids = [23523, 23526, 23559, 23592]
datastore.info()
obs_list = datastore.obs_list(obs_ids)


# RA and DEC of the Crab
RA_OBJ = 83.63
DEC_OBJ = 22.01

target_position = SkyCoord(ra=RA_OBJ, dec=DEC_OBJ, unit='deg', frame='icrs') # coordinates of the Crab 
on_region_radius = Angle('0.11 deg')
on_region = CircleSkyRegion(center=target_position, radius=on_region_radius)


EXCLUSION_FILE = '$GAMMAPY_EXTRA/datasets/exclusion_masks/tevcat_exclusion.fits'

allsky_mask = SkyImage.read(EXCLUSION_FILE)

exclusion_mask = allsky_mask.cutout(
    position=on_region.center,
    size=Angle('6 deg'),
)


# background estimator
background_estimator = background.ReflectedRegionsBackgroundEstimator(
    obs_list=obs_list,
    on_region=on_region,
    exclusion_mask = exclusion_mask)

background_estimator.run()
print background_estimator.result[0]

plt.figure(figsize=(6,6))
background_estimator.plot()
plt.show()

stats = []
for obs, bkg in zip(obs_list, background_estimator.result):
    stats.append(data.ObservationStats.from_obs(obs, bkg))
    
obs_summary = data.ObservationSummary(stats)
fig = plt.figure(figsize=(10,6))
ax1=fig.add_subplot(121)

obs_summary.plot_excess_vs_livetime(ax=ax1)
ax2=fig.add_subplot(122)
obs_summary.plot_significance_vs_livetime(ax=ax2)


#  spectrum extraction
extraction = SpectrumExtraction(obs_list=obs_list,
                                bkg_estimate=background_estimator.result,
                                containment_correction=False
                                )

extraction.run()

# diagnostic plots
extraction.observations[0].peek()
plt.show()

# check the energy dispersion
edisp =  extraction.observations[0].edisp
etru = np.sqrt(edisp.e_true.bins[:-1] *  edisp.e_true.bins[1:])
plt.plot(etru, edisp.get_bias(etru))
plt.xscale('log')
plt.xlabel('E [TeV]')
plt.ylabel('Bias')
plt.show()

edisp.plot_bias()
plt.show()
# check on the dispersion
edisp.get_resolution(etru)
```

**Comment by cdeil** [comment]
@cosimoNigro - Thanks for finding and filing the issue!

There's no need to run a spectral analysis to trigger the issue. Here's a minimal example:
```python
import numpy as np
import astropy.units as u
from gammapy.irf import EnergyDispersion
e_true = np.logspace(-1, 2, 30) * u.TeV
e_reco = np.logspace(-2, 2, 20) * u.TeV
edisp = EnergyDispersion.from_gauss(e_true, e_reco)
edisp.get_resolution([1, 2] * u.TeV)
```
which will result in this error:
```
>>> edisp.get_resolution([1, 2] * u.TeV)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/deil/code/gammapy/gammapy/irf/energy_dispersion.py", line 348, in get_resolution
    temp = (self.e_reco._interp_nodes() - mean) ** 2
  File "/Users/deil/Library/Python/3.6/lib/python/site-packages/astropy/units/quantity.py", line 641, in __array_ufunc__
    *arrays, **kwargs)
ValueError: operands could not be broadcast together with shapes (19,) (2,) 
```

Now I guess this could be considered a bug report of feature request, because `edisp.get_resolution` does work for a single scalar energy, and the docstring isn't clear whether it's supposed to work for a 1-dim array of energies:
http://docs.gammapy.org/en/latest/api/gammapy.irf.EnergyDispersion.html#gammapy.irf.EnergyDispersion.get_resolution

@joleroi @lmohrmann - Thoughts?
I guess the options are to make it work, or to clarify in the docstring or raise a better error.
If anyone is willing to send a PR to make this better, please say so in the comment; otherwise I'll do it tomorrow before releasing Gammapy 0.7.

**Comment by cosimoNigro** [comment]
Hi @cdeil, sorry for the long code, I just modified what I was using. 
Getting back to your minimal example I have noticed also that if you choose a finer energy you have the issue you can see in the plot.
Both the method `get_bias` and `get_resolution` work exactly at `e_true` points, if you feed arbitrary value they start to have this behaviour, maybe this has to be clarified in the docs or modified.

```python
import numpy as np
import astropy.units as u
from gammapy.irf import EnergyDispersion
import matplotlib.pyplot as plt

e_true = np.logspace(-1, 2, 30) * u.TeV
e_reco = np.logspace(-2, 2, 20) * u.TeV
edisp = EnergyDispersion.from_gauss(e_true, e_reco)

ene_1 = [10, 20, 30, 40] * u.TeV
ene_2 = np.linspace(10,40,16) * u.TeV

reso_1 = [edisp.get_resolution(ene) for ene in ene_1]
reso_2 = [edisp.get_resolution(ene) for ene in ene_2]

plt.plot(ene_1, reso_1, lw=1.8, label='ene_1')
plt.plot(ene_2, reso_2, lw=1.8, label='ene_2')
plt.legend()
plt.xscale('log')
plt.xlabel('E [TeV]')
plt.ylabel('resolution')
plt.show()
```

**Comment by cosimoNigro** [comment]
Also `get_bias` has the same zig zag behaviour, for arbitrary points

```python
import numpy as np
import astropy.units as u
from gammapy.irf import EnergyDispersion
import matplotlib.pyplot as plt

e_true = np.logspace(-1, 2, 30) * u.TeV
e_reco = np.logspace(-2, 2, 20) * u.TeV
e_true_arbitrary = np.logspace(-1, 1, 15) * u.TeV

edisp = EnergyDispersion.from_gauss(e_true, e_reco)

bias_1 = edisp.get_bias(e_true)
bias_2 = edisp.get_bias(e_true[::2])
bias_3 = edisp.get_bias(e_true_arbitrary)

ax = edisp.plot_bias(label='from plot_bias', lw=1.8)
ax.plot(e_true, bias_1, lw=1.8, label='e_true')
ax.plot(e_true[::2], bias_2, lw=1.8, label='e_true[::2]', ls='--')
ax.plot(e_true_arbitrary, bias_3, lw=1.8, label='e_true_arbitrary', ls='--')
plt.legend()
plt.xscale('log')
plt.xlabel('E [TeV]')
plt.ylabel('bias')
plt.show()
```
![figure_1-3](https://user-images.githubusercontent.com/26409711/30428631-952be0f0-9954-11e7-8c91-e5ff4a743738.png)

**Comment by lmohrmann** [comment]
I think there are actually two issues here. One is that `get_resolution` cannot handle multi-dimensional inputs. I just created a PR that fixes that (see #1133). The other is the "zig-zag" behaviour that @cosimoNigro observes. I think that is likely due to the linear interpolation that is performed if the pdf is queried at energies between nodes (although the steps do seem quite large). I don't have a fix for this yet, but may look into this if I can find the time.

**Comment by joleroi** [comment]
> I think that is likely due to the linear interpolation that is performed if the pdf is queried at energies between nodes (although the steps do seem quite large). I don't have a fix for this yet, but may look into this if I can find the time.

I looked into this, and the problem is rather the opposite. There is **no** interpolation performed on the PDF (see [here](https://github.com/gammapy/gammapy/blob/master/gammapy/irf/energy_dispersion.py#L39) )

This was done in https://github.com/gammapy/gammapy/commit/468803b9a3cfd6782a53c8971b6bf8d713d9858f in https://github.com/gammapy/gammapy/pull/956 following a discussion in https://github.com/gammapy/gammapy/issues/882 and https://github.com/gammapy/gammapy/issues/953

In short: The idea was that all interpolation must happend on ``EnergyDispersion2D`` and ``EnergyDispersion`` is just a matrix for the spectral analysis.

The problem is now that in ``get_bias`` the mean reco energy does not change within a true energy bin, but the bias is compared with the true energy that is passed as an argument to ``get_bias``, thus the zig-zaggy behaviour. I propose to change ``get_bias`` and always return the bias in the bin the handed true energy falls into. If there are no objection I will file a PR

**Comment by joleroi** [comment]
@cosimoNigro What's the status here? Can you check if you still see the described problem after #1138

**Comment by cosimoNigro** [comment]
Yes @joleroi I still have the same zig-zag effect
Updating the snippet
```python
import numpy as np
import astropy.units as u
from gammapy.irf import EnergyDispersion
import matplotlib.pyplot as plt

e_true = np.logspace(-1, 2, 30) * u.TeV
e_reco = np.logspace(-2, 2, 20) * u.TeV
e_true_arbitrary = np.logspace(-1, 1, 15) * u.TeV
 
edisp = EnergyDispersion.from_gauss(e_true, e_reco, sigma=0.2, bias=0.)
 
bias_1 = edisp.get_bias(e_true)
bias_2 = edisp.get_bias(e_true[::2])
bias_3 = edisp.get_bias(e_true_arbitrary)
 
ax = edisp.plot_bias(label='from plot_bias', lw=1.8)
ax.plot(e_true, bias_1, lw=1.8, label='e_true')
ax.plot(e_true[::2], bias_2, lw=1.8, label='e_true[::2]', ls='--')
ax.plot(e_true_arbitrary, bias_3, lw=1.8, label='e_true_arbitrary', ls='--')
plt.legend()
plt.xscale('log')
plt.xlabel('E [TeV]')
plt.ylabel('bias')
plt.show()
```
(just modified `from_gauss` wrt the previous snippet). 
Try it yourself and let me know.
![figure_1](https://user-images.githubusercontent.com/26409711/37669735-df873c48-2c67-11e8-9d04-042f7fe869ff.png)

**Comment by cdeil** [comment]
@joleroi - Thoughts?

**Comment by joleroi** [comment]
Well, the issue lies in the finite resolution of the ``e_reco`` histogram in each bin of true energy. I modified @cosimoNigro script a little bit to illustrate this

```
import astropy.units as u
from gammapy.irf import EnergyDispersion
import matplotlib.pyplot as plt

e_true = np.logspace(-1, 2, 30) * u.TeV
e_reco = np.logspace(-2, 2, 20) * u.TeV
test_e = np.logspace(-1, 2, 1000) * u.TeV

edisp = EnergyDispersion.from_gauss(e_true, e_reco, sigma=0.2, bias=0.)
bias = edisp.get_bias(test_e)
plt.plot(test_e, bias)
plt.semilogx()
```
![edisp_issue](https://user-images.githubusercontent.com/11678435/41973727-24d908e8-7a16-11e8-9e87-d6f2868827e5.png)

after changing the reco energy binning from 20 to 200 bins the problem is much less severe

![edisp_issue2](https://user-images.githubusercontent.com/11678435/41973750-345a8be8-7a16-11e8-9a03-6b412132ad87.png)

I don't know how to proceed here

**Comment by cdeil** [comment]
@cosimoNigro - What to you suggest? Some change to the implementation, or docstring clarification concerning the method that's used for the computation?

**Comment by cosimoNigro** [comment]
Ok, I devoted a couple of hours and my opinion is: it is fine as it is.
What @joleroi is doing in the function to get the bias is: 
1) compute the mean of `e_reco` in the node of `e_true` to which the input energy is close
https://github.com/gammapy/gammapy/blob/master/gammapy/irf/energy_dispersion.py#L366
2) get the `e_true` value of the node closer to the input energy
https://github.com/gammapy/gammapy/blob/master/gammapy/irf/energy_dispersion.py#L368
3) compute the bias following the formula

I'll add another example to the discussion: let us define a migration matrix this time with a finer energy binning in `e_reco` (this is what we do usually in MAGIC) and a `test_e` array that is denser than `e_true` and that we will use to compute the bias
```python
import numpy as np
import astropy.units as u
from gammapy.irf import EnergyDispersion
from gammapy.utils.energy import EnergyBounds
import matplotlib.pyplot as plt

e_true = np.logspace(-1, 2, 10) * u.TeV
e_reco = np.logspace(-1.5, 2.5, 20) * u.TeV
test_e = np.logspace(-0.5, 1.5, 20)* u.TeV

edisp = EnergyDispersion.from_gauss(e_true, e_reco, sigma=0.3, bias=0.)
fig,ax = plt.subplots()
edisp.plot_matrix(ax=ax)
ax.plot(test_e, edisp.get_mean(test_e), marker='o', ls='', color='crimson', label='get_mean()')
ax.legend()
```
I plot the computed mean for `e_reco` (that  is subtracted to `e_true`) on top of the migration matrix. It seems to me that the values match with the denser part of the matrix in each bin, i.e. the `e_true` mean is properly estimated and point 1. is fine.
![index](https://user-images.githubusercontent.com/26409711/42032373-ab995cfc-7ad9-11e8-82a0-d07380409967.png)


If we plot the `get_bias` result on `test_e` we see the fluctuations. They are a result of the finite binning I think
```python
bias = edisp.get_bias(test_e)
plt.plot(test_e, bias, marker='o', ls='-')
plt.semilogx()
plt.xlabel('E$_{true}$ / TeV')
plt.show()
```
![index](https://user-images.githubusercontent.com/26409711/42032623-7608ae02-7ada-11e8-858d-0bb430c61163.png)
Furthermore you can see that the range of variation is really small, second decimal figure fluctuations.
If we specify a plot range compatible with the sigma / resolution of the migration matrix
```python
bias = edisp.get_bias(test_e)
plt.plot(test_e, bias, marker='o', ls='-')
plt.semilogx()
plt.ylim(-0.3,0.3)
plt.xlabel('E$_{true}$ / TeV')
plt.show()
``` 
we don't see strong variations. 
![index](https://user-images.githubusercontent.com/26409711/42032833-1f12c104-7adb-11e8-8ecd-00c30f3e110f.png)

So I am convinced the way it is done now it's fine. If one takes into account the proper scale in the plot and if one has a fine-binned `e_reco` matrix the problem should not appear.

I just have  a question to @joleroi. 
Are IRFs interpolated in `gammapy`? 
I am not sure I understand why a specified value of `e_true` the index number for nodes and bins can be different (continuing the previous snippet)
```python
e_true_bounds = EnergyBounds(e_true)
print(edisp.e_true.find_node(test_e))
```
```
[1 1 2 2 2 2 3 3 3 4 4 4 5 5 5 6 6 6 7 7]
```
```python
print(e_true_bounds.find_energy_bin(test_e))
```
```
[1 1 2 2 2 3 3 3 4 4 4 4 5 5 5 6 6 6 7 7]
```
but in principle I should get the same values no?
Would it change the results doing the calculations with the values **contained** in one bin wrt the ones **closer** to a node?

Thanks

**Comment by cdeil** [comment]
Discussed in the Gammapy dev call just now with @cosimoNigro .
Basically everything is OK.

If someone looks at EDISP again in the future, please start by trying to use Cosimo's test case from above to derive a better test case than what we currently have:

https://github.com/gammapy/gammapy/blob/377ad3e657cbe4a9663e45b96d7aa6ce276f6bfb/gammapy/irf/tests/test_energy_dispersion.py#L59
